### ISSUE TITLE
Phase 51.2: CrossModalAligner — Contrastive Alignment & Cross-Modal Retrieval

### ISSUE THREAD

**Issue by web3guru888** [web3guru888/asi-build#987]
### Overview
Implement contrastive alignment mechanisms (CLIP-style) for creating unified multi-modal embedding spaces, enabling cross-modal retrieval, zero-shot classification, and semantic similarity across modalities.

### Architecture

```
┌──────────────────────────────────────────────────────────────────┐
│                     CrossModalAligner                            │
│                                                                  │
│  ┌──────────────┐        ┌──────────────────────┐               │
│  │  Modality A   │───────▶│  Contrastive Loss    │               │
│  │  Embedding    │        │  ┌────────────────┐  │               │
│  └──────────────┘        │  │ InfoNCE / NT-   │  │               │
│                          │  │ Xent / SigLIP   │  │               │
│  ┌──────────────┐        │  └────────────────┘  │               │
│  │  Modality B   │───────▶│  ┌────────────────┐  │               │
│  │  Embedding    │        │  │ Hard Negative   │  │               │
│  └──────────────┘        │  │ Mining          │  │               │
│                          │  └────────────────┘  │               │
│  ┌──────────────┐        │  ┌────────────────┐  │               │
│  │  Modality C   │───────▶│  │ Temperature    │  │               │
│  │  Embedding    │        │  │ Scheduling     │  │               │
│  └──────────────┘        │  └────────────────┘  │               │
│                          └──────────────────────┘               │
│                                    │                             │
│                          ┌─────────▼──────────┐                 │
│                          │  Aligned Embedding  │                 │
│                          │  Space              │                 │
│                          └────────────────────┘                 │
└──────────────────────────────────────────────────────────────────┘
```

### Key Components

1. **ContrastiveLoss**: InfoNCE, NT-Xent, SigLIP sigmoid loss, and symmetric cross-entropy variants
2. **HardNegativeMiner**: Mining strategies — in-batch negatives, semi-hard mining, MoCo-style queue
3. **TemperatureScheduler**: Learnable temperature parameter with scheduling (CLIP-style log-parameterized)
4. **PairSampler**: Efficient positive/negative pair sampling across modalities
5. **AlignmentScorer**: Cosine similarity, bilinear scoring, and learned distance metrics
6. **RetrievalIndex**: FAISS/Annoy-based approximate nearest neighbor index for fast cross-modal retrieval
7. **ZeroShotClassifier**: CLIP-style zero-shot classification using text prompts as class prototypes

### Interface

```typescript
interface ICrossModalAligner {
  align(embeddings: Map<ModalityType, LatentEmbedding[]>): Promise<AlignmentResult>;
  computeSimilarity(a: LatentEmbedding, b: LatentEmbedding): number;
  retrieve(query: LatentEmbedding, targetModality: ModalityType, topK: number): Promise<RetrievalResult[]>;
  zeroShotClassify(input: LatentEmbedding, classPrompts: string[]): Promise<ClassificationResult>;
  buildIndex(embeddings: LatentEmbedding[], modality: ModalityType): Promise<void>;
  getAlignmentScore(modalityA: ModalityType, modalityB: ModalityType): number;
}

interface AlignmentResult {
  loss: number;
  pairwiseScores: Map<string, number>;
  temperature: number;
  hardNegativeStats: NegativeStats;
}

interface RetrievalResult {
  embedding: LatentEmbedding;
  score: number;
  rank: number;
  metadata: Record<string, unknown>;
}
```

### Implementation Details

- **Symmetric Contrastive Learning**: Loss computed in both directions (A→B and B→A)
- **Multi-Modal Alignment**: Extend beyond pairs to N-way alignment (ImageBind-style binding)
- **Momentum Encoder**: Optional MoCo-style momentum encoder for large effective batch sizes
- **Gradient Caching**: GradCache-style gradient accumulation for memory-efficient training
- **Evaluation**: R@1, R@5, R@10 recall metrics for retrieval evaluation

### Acceptance Criteria

- [ ] InfoNCE and NT-Xent losses implemented with gradient correctness verified
- [ ] Hard negative mining improves retrieval recall by ≥5% vs random negatives
- [ ] Temperature parameter is learnable and stabilizes during training
- [ ] Cross-modal retrieval (text↔image) achieves R@10 ≥ 80% on validation set
- [ ] Zero-shot classification works with arbitrary text prompts
- [ ] FAISS index supports >1M embeddings with <10ms query time
- [ ] Unit tests with ≥90% coverage
- [ ] Supports 2-way and N-way alignment configurations

### Dependencies
- Phase 51.1 (ModalityEncoder) — provides input embeddings
- Phase 46 (Self-Supervised Learning) — contrastive learning foundations

### References
- Radford et al. (2021) — CLIP contrastive learning
- Chen et al. (2020) — SimCLR / NT-Xent loss
- He et al. (2020) — MoCo momentum contrast
- Girdhar et al. (2023) — ImageBind multi-modal binding
- Zhai et al. (2023) — SigLIP sigmoid loss

*Part of [Phase 51 Planning](#985)*
